Makefile

```
@@ -37,51 +40,55 @@ base: $(PREFIX_PROG_STRIPPED)plo-$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).elf  $(PRE
 
 
-$(BUILD_DIR)/script.plo:
-	@printf "TOUCH script.plo\n"
-	$(SIL)touch $(BUILD_DIR)/script.plo
+$(BUILD_DIR)/script.plo $(BUILD_DIR)/ramscript.plo:
+	@echo "TOUCH $(@F)"
+	$(SIL)touch $@
 
 
 $(PREFIX_O)/script.o.plo: $(PREFIX_O)cmds/cmd.o $(BUILD_DIR)/script.plo
 	@mkdir -p $(@D)
-	@printf "EMBED script.plo\n"
-	$(SIL)$(OBJCOPY) --update-section .data=$(BUILD_DIR)/script.plo $(PREFIX_O)cmds/cmd.o --add-symbol script=.data:0 $(PREFIX_O)script.o.plo
-
-
-$(BUILD_DIR)/ramscript.plo:
-	@printf "TOUCH script.plo\n"
-	$(SIL)touch $(BUILD_DIR)/ramscript.plo
+	@echo "EMBED script.plo"
+	$(SIL)$(OBJCOPY) --update-section .data=$(BUILD_DIR)/script.plo $(PREFIX_O)cmds/cmd.o --add-symbol script=.data:0 $@
 
 
 $(PREFIX_O)/ramscript.o.plo: $(PREFIX_O)cmds/cmd.o $(BUILD_DIR)/ramscript.plo
 	@mkdir -p $(@D)
-	@printf "EMBED ramscript.plo\n"
-	$(SIL)$(OBJCOPY) --update-section .data=$(BUILD_DIR)/ramscript.plo $(PREFIX_O)cmds/cmd.o --add-symbol script=.data:0 $(PREFIX_O)ramscript.o.plo
+	@echo "EMBED ramscript.plo"
+	$(SIL)$(OBJCOPY) --update-section .data=$(BUILD_DIR)/ramscript.plo $(PREFIX_O)cmds/cmd.o --add-symbol script=.data:0 $@
 
 
-$(PREFIX_PROG)plo-$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).elf: $(OBJS) $(PREFIX_O)/script.o.plo
+$(PREFIX_PROG)plo-$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).elf: $(PREFIX_O)/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).ld $(OBJS) $(PREFIX_O)/script.o.plo
 	@mkdir -p $(@D)
-	@(printf "LD  %-24s\n" "$(@F)");
-	$(SIL)$(LD) $(LDFLAGS) -e _start --section-start .init=$(INIT_FLASH) $(BSS) -o $(PREFIX_PROG)plo-$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).elf $(OBJS) $(PREFIX_O)/script.o.plo $(GCCLIB)
+	@echo "LD  $(@F)"
+	$(SIL)$(LD) $(LDFLAGS) -Map=$<.map -o $@ -T $^ $(GCCLIB)
 
 
-$(PREFIX_PROG)plo-ram-$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).elf: $(OBJS) $(PREFIX_O)/ramscript.o.plo
+$(PREFIX_PROG)plo-ram-$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).elf: $(PREFIX_O)/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY)-ram.ld $(OBJS) $(PREFIX_O)/ramscript.o.plo
 	@mkdir -p $(@D)
-	@(printf "LD  %-24s\n" "$(@F)");
-	$(SIL)$(LD) $(LDFLAGS) -e _start --section-start .init=$(INIT_RAM) $(BSS) -o $(PREFIX_PROG)plo-ram-$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).elf $(OBJS) $(PREFIX_O)/ramscript.o.plo $(GCCLIB)
+	@echo "LD  $(@F)"
+	$(SIL)$(LD) $(LDFLAGS) -Map=$<.map -o $@ -T $^ $(GCCLIB)
 
 
 $(PREFIX_PROG_STRIPPED)%.hex: $(PREFIX_PROG_STRIPPED)%.elf
-	@(printf "HEX %s\n" "$(@F)");
+	@echo "HEX $(@F)"
 	$(SIL)$(OBJCOPY) -O ihex $< $@
 
 
 $(PREFIX_PROG_STRIPPED)%.img: $(PREFIX_PROG_STRIPPED)%.elf
-	@(printf "BIN %s\n" "$(@F)");
+	@echo "BIN $(@F)"
 	$(SIL)$(OBJCOPY) -O binary $< $@
 
 
-.PHONY: clean
+$(PREFIX_O)/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).ld:
+	@echo "GEN $(@F)"
+	$(SIL)$(CC) $(LDSFLAGS) -D__LINKER__ -undef -xc -E -P ld/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).ldt > $@
+
+
+$(PREFIX_O)/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY)-%.ld:
+	@echo "GEN $(@F)"
+	$(SIL)$(CC) $(LDSFLAGS) -D__LINKER__ -D$* -undef -xc -E -P ld/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).ldt > $@
+
```

```suggestion
LD_FILES := ld/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).ldt
ifeq ($(findstring  arm,$(TARGET_FAMILY)),arm)
  LD_FILES += ld/common/plo-arm.lds
else ifeq ($(findstring  ia32,$(TARGET_FAMILY)),ia32)
  LD_FILES += ld/common/plo-ia32.lds
endif


$(PREFIX_O)/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).ld: $(LD_FILES)
	@echo "GEN $(@F)"
	$(SIL)$(CC) $(LDSFLAGS) -D__LINKER__ -undef -xc -E -P $< > $@


$(PREFIX_O)/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY)-%.ld: $(LD_FILES)
	@echo "GEN $(@F)"
	$(SIL)$(CC) $(LDSFLAGS) -D__LINKER__ -D$* -undef -xc -E -P $< > $@

```
Changing the `*.ldt` or `*.lds` file will create a `*.ld` file again.

---

I have verified Your suggestion and it works just fine, thanks. I appreciate Your support in this topic. However, for now, I prefer the PR approach and here is why. There is no need to extend the `ifeq`, `LD_FILES` list with the complexity of the future hierarchy defined in the `*.ldt`, by inclusion of `*.lds` files, this should work like a simple template system, take top (target) file and parse the hierarchy (using C preprocessor) of the included sub-scripts, which may differ per target. E.g. :
* `imxrt`'s can soon be extended with `FLEXRAM_CONFIG` bitmap configuration scripts based on `ocram/dtcm/itcm` maps (for the `imxrt10xx` family will be different than for `imxrt117x`).
* some common parts of `plo-ia32.lds` and `plo-arm.lds` will be shared in separate sub-scripts (I believe there will be also `plo-riscv64.lds` in near future).
* placement of the device tree

The suggested change assumes that every time someone add a sub-script, he/she would have to modify the `LD_FILES` list with these newly added sub-scripts. There is effectively one `ld`, which is the processed output, one input `ldt` (top of target), and more `lds` (sub target) are mixed `linker`, `C`, and `assembler` partials.

I would accept the addition of (from the suggestion):
```Makefile
LD_FILES := ld/$(TARGET_FAMILY)-$(TARGET_SUBFAMILY).ldt
```
but, someone later may have mistakenly extended this list. The preferred way is to use `#include` to import sub scripts in the top target or sub target files.

Although, the part of `Makefile` responsible for `ld` file generation may be moved to `ld/` subdirectory, but not in this PR.

About the optimizations of the `plo` size, `hal_memoryGetNextEntry()` may be optimized (I agree), but other parts too of course not in this PR. I wanted the PR to be as simple as possible to easily switch to the `linker` scripts because of upcoming FlexSPI setup on `imxrt`'s require relocation of some parts of code to the `itcm` RAM (for the `.noxip` or `.ramfunc` section), which is the priority now.